src/shadow-embed.js

```diff
@@ -47,3 +125,91 @@ export function createShadowEmbedRoot(hostElement, extensionIds) {
 
   return shadowRoot;
 }
+
+
+/**
+ * Imports a body into a shadow root with the workaround for a unpolyfilled.
+ * @param {!ShadowRoot} shadowRoot
+ * @param {!Element} body
+ * @return {!Element}
+ */
+export function importShadowBody(shadowRoot, body) {
+  const doc = shadowRoot.ownerDocument;
+  let resultBody;
+  if (isShadowDomSupported()) {
+    resultBody = doc.importNode(body, true);
+  } else {
+    resultBody = doc.createElement('amp-body');
+    for (let n = body.firstChild; !!n; n = n.nextSibling) {
+      resultBody.appendChild(doc.importNode(n, true));
+    }
+    resultBody.style.display = 'block';
+  }
+  resultBody.style.position = 'relative';
+  shadowRoot.appendChild(resultBody);
+  return resultBody;
+}
+
+
+/**
+ * Adds the given css text to the given shadow root.
+ *
+ * The style tags will be at the beginning of the shadow root before all author
+ * styles. One element can be the main runtime CSS. This is guaranteed
+ * to always be the first stylesheet in the doc.
+ *
+ * @param {!ShadowRoot} shadowRoot
+ * @param {string} cssText
+ * @param {boolean=} opt_isRuntimeCss If true, this style tag will be inserted
+ *     as the first element in head and all style elements will be positioned
+ *     after.
+ * @param {string=} opt_ext
+ */
+export function installStylesForShadowRoot(shadowRoot, cssText,
+    opt_isRuntimeCss, opt_ext) {
+  insertStyleElement(
+      shadowRoot.ownerDocument,
+      shadowRoot,
+      transformShadowCss(cssText),
+      opt_isRuntimeCss || false,
+      opt_ext || null);
+}
+
+
+/**
+ * Copies runtime styles from the ampdoc context into a shadow root.
+ * @param {!./service/ampdoc-impl.AmpDoc} ampdoc
+ * @param {!ShadowRoot} shadowRoot
+ */
+export function copyRuntimeStylesToShadowRoot(ampdoc, shadowRoot) {
+  const style = dev().assert(
+      ampdoc.getRootNode().querySelector('style[amp-runtime]'),
+      'Runtime style is not found in the ampdoc: %s', ampdoc.getRootNode());
+  const cssText = style.textContent;
+  installStylesForShadowRoot(shadowRoot, cssText, /* opt_isRuntimeCss */ true);
+}
+
+
+/**
+ * Transforms CSS to isolate AMP CSS within the shaodw root and reduce the
+ * possibility of high-level conflicts.
+ * @param {string} css
+ * @return {string}
+ */
+export function transformShadowCss(css) {
+  if (isShadowDomSupported()) {
+    return css;
+  }
+
+  // Replace html/body with amp-html/amp-body.
+  const res = css.replace(/(html|body)/g, function(match, name, pos) {
```

Comment:
Yeah, we could totally do that. I guess there are degrees of isolation we can add. I thought this is an ok start and add more in follow up PRs. I'm currently investigating `<style scoped>` and `:scope` that could potentially help in this.